File: src/content/algorithms/dfs/code.py
def dfs_recursive(graph, start, visited=None):  # #@function-def
    if visited is None:  # #@init-visited
        visited = set()  # #@init-visited

    if start in visited:  # #@base-case
        return  # #@base-case

    visited.add(start)  # #@mark-visited
    print(start)  # #@process-node

    for neighbor in graph[start]:  # #@recurse
        dfs_recursive(graph, neighbor, visited)  # #@recurse


def dfs_iterative(graph, start):  # #@function-def
    visited = set()  # #@init-visited
    stack = [start]  # #@stack-push

    while stack:  # #@main-loop
        vertex = stack.pop()  # #@stack-pop

        if vertex in visited:  # #@base-case
            continue  # #@base-case

        visited.add(vertex)  # #@mark-visited
        print(vertex)  # #@process-node

        for neighbor in graph[vertex]:  # #@recurse
            if neighbor not in visited:  # #@recurse
                stack.append(neighbor)  # #@stack-push

File: src/content/algorithms/dfs/code.rs
use std::collections::{HashMap, HashSet};  // #@init-visited

pub fn dfs_recursive(  // #@function-def
    graph: &HashMap<char, Vec<char>>,
    start: char,
    visited: &mut HashSet<char>,
) {
    if visited.contains(&start) {  // #@base-case
        return;
    }

    visited.insert(start);  // #@mark-visited
    println!("{}", start);  // #@process-node

    if let Some(neighbors) = graph.get(&start) {  // #@recurse
        for &neighbor in neighbors {
            dfs_recursive(graph, neighbor, visited);  // #@recurse
        }
    }
}

pub fn dfs_iterative(  // #@function-def
    graph: &HashMap<char, Vec<char>>,
    start: char,
) -> Vec<char> {
    let mut visited = HashSet::new();  // #@init-visited
    let mut stack = vec![start];  // #@stack-push
    let mut result = Vec::new();

    while let Some(vertex) = stack.pop() {  // #@main-loop  // #@stack-pop
        if visited.contains(&vertex) {  // #@base-case
            continue;
        }

        visited.insert(vertex);  // #@mark-visited
        result.push(vertex);  // #@process-node

        if let Some(neighbors) = graph.get(&vertex) {  // #@recurse
            for &neighbor in neighbors.iter().rev() {
                if !visited.contains(&neighbor) {
                    stack.push(neighbor);  // #@stack-push
                }
            }
        }
    }

    result
}
